clusterloader2/pkg/measurement/common/network/network_performance_measurement.go

```diff
@@ -0,0 +1,485 @@
+	if npm.framework == nil {
+		klog.V(1).Infof("Network measurement %s wasn't started, skipping the Dispose() step", npm)
+		return
+	}
+	klog.Infof("Stopping network performance measurement...")
+	if err := npm.deleteCRD(); err != nil {
+		klog.Errorf("error while deleting CRD: %v", err)
+	}
+
+	if err := npm.deleteClusterRoleBinding(); err != nil {
+		klog.Errorf("error while deleting rbac: %v", err)
+	}
+	if err := client.DeleteNamespace(npm.k8sClient, netperfNamespace); err != nil {
+		klog.Errorf("error while deleting %s namespace: %v", netperfNamespace, err)
+	}
+	if err := client.WaitForDeleteNamespace(npm.k8sClient, netperfNamespace); err != nil {
+		klog.Errorf("error while waiting for %s namespace to be deleted: %v", netperfNamespace, err)
+	}
+
+	npm.closeCh()
+}
+
+func (npm *networkPerformanceMeasurement) watchNetworkTestEvents() {
+	dynamicClient := npm.framework.GetDynamicClients().GetClient()
+	gvk := schema.GroupVersionKind{Group: customResourceGroup, Kind: customResourceKind, Version: customResourceVersion}
+	gvr, _ := meta.UnsafeGuessKindToResource(gvk)
+	watchResource, err := dynamicClient.Resource(gvr).Watch(context.TODO(), metav1.ListOptions{})
+	if err != nil {
+		klog.Error("Error watching", err)
+		return
+	}
+	eventChannel := watchResource.ResultChan()
+	for event := range eventChannel {
+		var metricResp metricResponse
+		if event.Type != watch.Modified {
+			continue
+		}
+		unstructuredList, _ := event.Object.(runtime.Unstructured)
+		unstructuredObjContent := unstructuredList.UnstructuredContent()
+		metricMap := unstructuredObjContent["status"].(map[string]interface{})
+		getMetricResponse(metricMap, &metricResp)
+		if len(metricResp.Metrics) == 0 && metricResp.Error != "" {
+			klog.Error("error message from worker pod: ", metricResp.Error)
+			continue
+		}
+		npm.metricLock.Lock()
+		npm.metricVal = append(npm.metricVal, metricResp.Metrics)
+		npm.metricLock.Unlock()
+	}
+}
+
+func (npm *networkPerformanceMeasurement) execNToMTest() {
+	go npm.createCustomResourcePerUniquePodPair()
+
+	npm.formUniquePodPair()
+}
+
+func (npm *networkPerformanceMeasurement) createCustomResourcePerUniquePodPair() {
+	for CRIndex := 0; ; CRIndex++ {
+		uniqPodPair := <-npm.podPairCh
+		if uniqPodPair.isLastPodPair {
+			break
+		}
+		if npm.startTimeStampForTestExecution == 0 {
+			npm.startTimeStampForTestExecution = getStartTimeStampForPod()
+		}
+		npm.templateMapping = npm.populateTemplate(uniqPodPair, CRIndex)
+		if err := npm.framework.ApplyTemplatedManifests(networkTestRequestFilePath, npm.templateMapping); err != nil {
+			klog.Error(err)
+		}
+		npm.podPairNumber++
+	}
+}
+
+func (npm *networkPerformanceMeasurement) formUniquePodPair() {
+	var uniqPodPair uniquePodPair
+	lastPodPair := uniquePodPair{isLastPodPair: true}
+
+	var unUsedPodIndex = 0
+	for {
+		for workerNode, podList := range npm.workerPodList {
+			unUsedPod, err := getUnusedPod(&podList)
+			npm.workerPodList[workerNode] = podList
```

Comment:
That's a bit strange pattern.

Instead maybe create a helper type (which will be implemented as array), that simply implements a Pop() function.

i.e. sth like:

type podList []pod

func (l *podList) Pop() (*pod, error) {
  if len(l) == 0 {
    return nil, unsupported
  }
  result := l[len(l)-1]
  l = l[:len(l)-1]
  return result, nil
}
```